### DISCUSSION TITLE
Help getting started

### DISCUSSION THREAD

**Discussion by Trollstone** [mockoon/mockoon#1499] [Help]
HI,

first of all I have to confess, I ain't a developer ;)

From time to time I have to configure an application to exchange data with different REST and/or SCIM 2.0 endpoints, and sometimes the simulation/testing is quite hard. 

Now I found Mockoon, and it seems to offer all I need.
So far I got an endpoint for /Users defined, and I am receiving the response JSON defined for that endpoint in my application (or Postman)

So far so good, but I need to simulate different responses for the same endpoint.
For example, accessing /Users - without any query parameter or something, I want to receive an array of users.
And accessing the same endpoint with a query parameter like filter I want to receive only a subset of that array, maybe only one user!

I found, it is possible to add rules to the Routes, representing an endpoint, but after defining a rule for the Route /Users matching the query parameter filer = xyz, I am receiving the same response, no matter what value for the filter is provided in the request.

What I am doing wrong?

is it possible to define a Route for exactly GET  /Users and some others for GET /Users with different query prarameters?
Kind regards

Thorsten 

![Bildschirmfoto 2024-09-12 um 16 40 58](https://github.com/user-attachments/assets/eae4b2a0-f2ed-4a00-ab07-44b37e3e4171)

**Comment by 255kb** [comment]
Hi, 

>first of all I have to confess, I ain't a developer ;) 
It's not a sin ;)

First of all, regarding the rules, I suggest you follow [this tutorial](https://mockoon.com/tutorials/responses-and-rules/) and read a bit of [the docs](https://mockoon.com/docs/latest/route-responses/dynamic-rules/) to have a better idea on how it works. 
The tldr is: 
1. you create multiple responses with rules
2. if a set of rules match the response get served (if follows the order in the menu)
3. if none matches, the "default" response (blue flag) is served.

So you need at least two responses, a default one, and one when you provide params.

The second topic is the query params. They are not part of the route declaration, which means you can only have one `/users`. Based on what you described, if you want to follow REST principles, you would rather create two routes, `/users` and `/users/:id`, as getting user by their `id` would be considered a "resource" and part of the path, instead of being a query param filter. 

Now, your setup looks like a good candidate for the CRUD routes. They follow REST principles and let you create one route that will automatically allow you to get specific resources, filter, paginate, etc. You could even search by id. 
However, there are some constraints. We also have [docs](https://mockoon.com/docs/latest/api-endpoints/crud-routes/) and a [tutorial](https://mockoon.com/tutorials/create-full-rest-api-crud-routes/) to help you getting started with CRUD routes. 

I hope this helps. Let me know if you have more questions.
